bootsched.8
.if n .ds Q \&"
.if t .ds Q ``
.if n .ds U \&"
.if t .ds U ''
.TH "BOOTSCHED" 8 
.tr \&
.nr bi 0
.nr ll 0
.nr el 0
.de DS
..
.de DE
..
.de Pp
.ie \\n(ll>0 \{\
.ie \\n(bi=1 \{\
.nr bi 0
.if \\n(t\\n(ll=0 \{.IP \\(bu\}
.if \\n(t\\n(ll=1 \{.IP \\n+(e\\n(el.\}
.\}
.el .sp 
.\}
.el \{\
.ie \\nh=1 \{\
.LP
.nr h 0
.\}
.el .PP 
.\}
..
.SH NAME

.Pp
\fBbootsched\fP - tool for setting startup time of PowerMacs
.Pp
.SH SYNOPSIS

.Pp
\fBbootsched\fP \f(CR[\fP\fB-q\fP\f(CR]\fP \f(CR[\fP\fB-w\(br-e\(br-d DAY\fP\f(CR]\fP \fB-t HH:MM\fP
.Pp
\fBbootsched\fP \f(CR[\fP\fB-q\fP\f(CR]\fP \fB+value\f(CR[\fPmhd\f(CR]\fP\fP
.Pp
\fBbootsched -h\fP
.Pp
.SH DESCRIPTION

.Pp
\fBbootsched\fP
sets the scheduled boot time in the hardware PMU on OldWorld PowerMacs.  
When set the machine will automatically boot at the specified time.
When cleared the machine must be manually booted.
\fBbootsched\fP can not be used on NewWorld PowerMacs.
.Pp
\fBbootsched\fP sets the boot time for the next startup only.
For continued, unattended operation
\fBbootsched\fP
should be invoked during system startup.
.Pp
.SH OPTIONS

.Pp
.nr ll +1
.nr t\n(ll 2
.if \n(ll>1 .RS
.IP "\fB-w\fP"
.nr bi 1
.Pp
Set next power-up on a weekday
.IP "\fB-e\fP"
.nr bi 1
.Pp
Set next power-up on a weekend day
.IP "\fB-d DAY\fP"
.nr bi 1
.Pp
Set next power-up on to DAY (0..6). Day 0 represents Sunday
.IP "\fB-t HH:MM\fP"
.nr bi 1
.Pp
Set time of day to power up
.IP "\fB+value[mhd]\fP"
.nr bi 1
.Pp
Set next power up at a time in seconds, minutes, hours or days 
ahead of current time
.IP "\fB-q\fP"
.nr bi 1
.Pp
Quiet mode
.IP ""
.nr bi 1
.Pp
.IP "\fB-h\fP"
.nr bi 1
.Pp
Print usage information.
.IP "\fB-v\fP"
.nr bi 1
.Pp
verbose operation.
.Pp
.if \n(ll>1 .RE
.nr ll -1
.Pp
.SH SEE ALSO

.Pp
autoboot(8)
.Pp
.SH AUTHORS

.Pp
.DS
.sp 
.ft RR
.nf
Michael Schmitz <schmitz@biophys.uni-duesseldorf.de> 
.DE
.fi 
.ec
.ft P
.sp
.Pp
.Pp
\fBbootsched\fP is derived from 
Takeshi Oe's <toe@ppc.linux.or.jp>
\fBpmacpow\fP,
as modified by Kieth Keller <kkeller@speakeasy.net>
.Pp
.Pp


debian/changelog
pmac-utils (1.1.3-27+ports) unreleased; urgency=medium

  * Upload to unreleased for powerpc and ppc64.
  * Bump compat level to 9 in debian/compat.
  * Drop #define __USE_GNU from nwnvsetenv.c.
  * Fix syntax error in autoboot.sgml.

 -- John Paul Adrian Glaubitz <glaubitz@physik.fu-berlin.de>  Tue, 08 Jan 2019 17:48:12 +0300

pmac-utils (1.1.3-27) unstable; urgency=medium

  * Incremental packaging updates for Standards-Version: 3.9.7:
    - Record our source format (1.0) in debian/source/format.
    - Stop redundantly setting DH_COMPAT in debian/rules.
    - Add build-indep and build-arch targets to debian/rules.
    - Use dh_prep instead of dh_clean -k in debian/rules.
    - Add comment-only watch file to document upstream state.
    - Use ${misc:Depends} in debian/control for binaries.
    - Remake debian/copyright with a DEP-5 compatible format.

 -- Adam Conrad <adconrad@0c3.net>  Tue, 05 Apr 2016 12:48:28 -0600

pmac-utils (1.1.3-26) unstable; urgency=medium

  * Add myselt to Uploaders with Michael's consent.
  * Rename to pmac-utils, powerpc-utils is another project.
  * Stop building on little-endian arches (LP: #1398163)
  * Drop ancient and obsolete Conflicts/Replaces.

 -- Adam Conrad <adconrad@0c3.net>  Sun, 27 Mar 2016 13:42:07 -0600

powerpc-utils (1.1.3-25) unstable; urgency=medium

  * Update maintainer address. Closes: #762790.
  * Fix compiler warnings.

 -- Michael Schmitz <schmitzmic@gmail.com>  Wed, 29 Oct 2014 20:28:32 +1100

powerpc-utils (1.1.3-24.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Add ppc64el to the Architecture: field. Closes: #738140.

 -- Aurelien Jarno <aurel32@debian.org>  Sat, 20 Sep 2014 22:09:45 +0200

powerpc-utils (1.1.3-24) unstable; urgency=low

  * Fix fixed postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file with the correct data otherwise). 
    Closes: #528443.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 20:04:24 +1100

powerpc-utils (1.1.3-23) unstable; urgency=low

  * Fix broken -f option for fnset. Check for presence of ADB devices and bail 
    out on USB based machines.
    Closes: #474814.

  * Fix postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file otherwise). 
    Closes: #528443.

  * Update bootsched manpage to clarify that it should only be used on 
    Oldworld PowerMacs.
    Closes: #407295.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 19:34:24 +1100

powerpc-utils (1.1.3-22) unstable; urgency=low

  * Fix problem with nvram reads segfaulting on efika, reported by Robert
    Millan <rm@aybabtu.com> and patched by same.
    Closes: #431116.

  * Honor DEB_BUILD_OPTS=noopt, patch by Robert Millan <rm@aybabtu.com>.
    Closes: #431117.

  * Add Aurelien Gerome <ag@roxor.cx> as co-maintainer.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  4 Jul 2007 17:21:40 +0100

powerpc-utils (1.1.3-21) unstable; urgency=low

  * Fix problem with truncated nvram reads/writes, reported by Adam
    Lackorzynski <adam@os.inf.tu-dresden.de>.
    Closes: #413131.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  7 Mar 2007 15:57:34 +0100

powerpc-utils (1.1.3-20) unstable; urgency=low

  * Change section to 'utils' to match override file.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-19) unstable; urgency=low

  * The 'upload lost' release :-(

  * Incorporate newworld autoboot patch by Rich Johnson.

  * Fix description formatting (thanks to Nick Boullis <nboullis@debian.org>
    for the alert).
    Closes: #336972.

  * Add ppc64 arch (just in case biarch powerpc isn't enough).
    Closes: #360797.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-18) unstable; urgency=low

  * add -D_GNU_SOURCE to Makefile in order to help with 64 bit pointer
    size on ppc64.
    Closes: #326394.

  * Fix problem introduced in -12 (/etc/adjtime converted to symlink 
    into /var fs): Convert symlink to plain copy now.
    Closes: #332926.

  * Fix char signedness bug in nvsetvol (patch by Eugen Dedu
    <Eugen.Dedu@pu-pm.univ-fcomte.fr>.
    Closes: #334547

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 19 Oct 2005 18:07:34 +0200

powerpc-utils (1.1.3-17) unstable; urgency=low

  * Fix for iBook G4 failure to save nvram data (as manifest by failure to
    properly store the boot chime volume). Patch by Joerg Dorchain
    <joerg@dorchain.net>.
    Closes: #321655, #320566.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 08 Aug 2005 14:40:34 +0200

powerpc-utils (1.1.3-16) unstable; urgency=low

  * Brown Paper Bag release: Added bootsched util (for CUDA based Powermacs)
    the source for which I had buried for over three years now.

  * Finally, actually install the autoboot binary (reported missing by
    Herve Eychenne <rv@eychenne.org>).
    Closes: #317021.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 06 Jul 2005 09:42:44 +0200

powerpc-utils (1.1.3-15) unstable; urgency=low

  * Merge back patches from ubuntu (thanks to Adam Conrad <adconrad@0c3.net>
    for the hint): fix linuxdoc-tools b-d, drop adb_mouse.h header. Update 
    description to include autoboot and lsprop tools.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 09 May 2005 12:31:24 +0100

powerpc-utils (1.1.3-14) unstable; urgency=low

  * Fixed nvsetvol failure (infinite loop reading NVRAM) on OldWorld
    machines (reported by Christophe Donatsch <cdonatsch@tele2.ch>).
    Closes: #304493.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Sun, 01 May 2005 11:58:37 +0100

powerpc-utils (1.1.3-13) unstable; urgency=low

  * Fixed nvsetenv's devfsd bug (reported by Roger Leigh <rleigh@debian.org>).
    Closes: #300946.

  * Added autoboot util (for CUDA based Powermacs)


bootsched.c
/*
 * bootsched.c - set the next power up time on VIA-CUDA.
 *
 * Debian version of pmacpow.c (essentially, renamed and docs added)
 * 2005-07-06 MSch
 *
 * Copyright (C) 1999 Takashi Oe.  All rights reserved.
 * Parts of the code are from clock.c and bat.c
 * by Paul Mackerras.
 *
 * Modified 2002 Keith Keller.
 * 
 * To make: save file to pmacpow.c
 * gcc -o pmacpow pmacpow.c
 * strip pmacpow
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#include "autoboot-conf.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <string.h>
#include <fcntl.h>
#include <getopt.h>
#include <time.h>
#include <sys/time.h>

#ifdef HAVE_ASM_ADB_H
# include <asm/adb.h>
# else
#  ifdef HAVE_LINUX_ADB_H
#   include <linux/adb.h>
#  endif
#endif

#ifdef HAVE_ASM_CUDA_H
# include <asm/cuda.h>
# else
#  ifdef HAVE_LINUX_CUDA_H
#   include <linux/cuda.h>
#  endif
#endif

unsigned char reqt[2] =
	{ CUDA_PACKET, CUDA_GET_TIME};
unsigned char reqp[6] =
	{ CUDA_PACKET, CUDA_POWERUP_TIME, 0, 0, 0, 0};

unsigned char reply[16];

/* RTC on PowerMacs stores seconds since 1 Jan 1904 */
#define RTC_OFFSET	2082844800

static int week = 0;
static int day = -1;
static int abstime = 0;
static int quiet = 0;
static char *program_name;

void dump(unsigned char *buf, int n)
{
	int i;

	for (i = 0; i < n; ++i)
		fprintf(stderr, " %.2x", buf[i]);
	fprintf(stderr, "\n");
}

void usage(void) {
	fprintf(stderr,
		"%s [-q] [-w|-e|-d DAY] -t HH:MM\n"
		"%s [-q] +value[mhd]\n"
		"\t-w: set next power-up on a weekday\n"
		"\t-e: set next power-up on a weekend day\n"
		"\t-d DAY: set next power-up to DAY (0..6);"
			" 0 represents Sunday\n"
		"\t-q: quiet mode\n"
		"\t-t HH:MM: time of day to power-up\n"
		"\t+value[mhd]: time in secs, mins, hours, or days ahead of"
			" current time\n",
		program_name, program_name);
	exit(1);
}

int next_wday(int wday) {
	if (wday == 0)
		return 1;
	if (wday == 6)
		return 2;
	return 0;
}

int next_weday(int wday) {
	if ((wday >=1) && (wday < 6))
		return 6 - wday;
	return 0;
}

int set_powerup_time(int fd, long up)
{
	int n;

	*(long *)(reqp+2) = up;
	n = write(fd, reqp, sizeof(reqp));
	if (n != sizeof(reqp)) {
		fprintf(stderr, "%s: write returned %d\n", program_name, n);
		return -1;
	}
	if ((n = read(fd, reply, sizeof(reply))) < 0) {
		perror("read");
		return -1;
	}
	if (n != 3) {
		dump(reply, n);
		fprintf(stderr, "%s: command failed\n", program_name);
		return -1;
	}
	return 0;
}

long get_current_time(int fd)
{
	int n;

	n = write(fd, reqt, sizeof(reqt));
	if (n != sizeof(reqt)) {
		fprintf(stderr, "%s: write returned %d\n", program_name, n);
		return -1;
	}
	if ((n = read(fd, reply, sizeof(reply))) < 0) {
		perror("read");
		return -1;
	}
	if (n != 7) {
		dump(reply, n);
		fprintf(stderr, "%s: command failed\n", program_name);
		return -1;
	}
	return (long) ((reply[3] << 24) + (reply[4] << 16)
			+ (reply[5] << 8) + (long) reply[6]);
}

long calc_powerup_time(int hour, int min)
{
	struct tm tm, ctm;
	time_t systime;

	systime = time(NULL);
	ctm = *localtime(&systime);
	tm = ctm;
	if (!quiet)
		fprintf(stderr, "current time is %s", asctime(&tm));
	tm.tm_sec = 0;
	tm.tm_min = min;
	tm.tm_hour = hour;
	if (mktime(&tm) < (systime + 60)) {
		++tm.tm_mday;
		++tm.tm_wday;
		(void)mktime(&tm);
	}
	if (day == -1) {
		if (week == 1)
			tm.tm_mday += next_wday(tm.tm_wday);
		if (week == 2)
			tm.tm_mday += next_weday(tm.tm_wday);
	} else {
		if (day != tm.tm_wday) {
			if (day < tm.tm_wday)
				day += 7;
			tm.tm_mday += day - tm.tm_wday;
		}
	}
	return (long)(mktime(&tm) + RTC_OFFSET);
}

int main(int argc, char **argv)
{
	int fd, arg, hour = 0, min = 0;
	long sec, uptime;
	char *timestr = NULL, *tmp;
	extern int optind;

	program_name = (char *)basename(argv[0]);
	if (getuid()) {
		fprintf(stderr, "Sorry, must be root to set power up time\n");
		return 1;
	}
	while ((arg = getopt(argc, argv, "wed:tq")) != EOF) {
		switch (arg) {
		case 0:
			break;
		case 'w':
			week = 1;
			break;
		case 'e':
			week = 2;
			break;
		case 'd':
			day = atoi(optarg);
			if ((day < 0) || (day > 6))
				usage();
			break;
		case 't':
			abstime = 1;
			break;
		case 'q':
			quiet = 1;
			break;
		default:
			usage();
		}
	}
	if (argv[optind])
		timestr = argv[optind];
	else
		usage();
	tmp = strsep(&timestr, ":");
	if (tmp && abstime) {
		if (!tmp || !timestr)
			usage();
		hour = atoi(tmp);
		min = atoi(timestr);
		sec = 0;
	} else {
		timestr = argv[optind];
		sec = strtol(timestr, &tmp, 10);
		if (tmp) {
			switch (tolower(tmp[0])) {
			case 'd':
				sec *= 24;
			case 'h':
				sec *= 60;
			case 'm':
				sec *= 60;
				break;
			}
		}
	}

	if ((fd = open("/dev/adb", O_RDWR)) < 0) {
		perror("open");
		return 1;
	}
	if (sec > 0) {
		long cur = get_current_time(fd);

		if (cur == -1)
			return 1;
		uptime = cur + sec;
	} else
		uptime = calc_powerup_time(hour, min);
	if (set_powerup_time(fd, uptime) < 0)
		return 1;
	close(fd);
	if (!quiet) {
		uptime -= RTC_OFFSET;
		fprintf(stderr, "will be up again %s",
			asctime(localtime(&uptime)));
	}
	return 0;
}
